### PR TITLE
Pull request kurnia

### PR DESCRIPTION
**Pull request by kurniaw**
Hi Nazmy,
Here are my submission. i haven't got time to complete task [MOKA-004] as i'm going to travel over this weekend. 
Thank you.

Regards,
Kurnia

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -1,11 +1,41 @@
...
+
+    def edit
+        @user = current_user
+    end
+
+    def create
+        @user = User.create(user_params)
+
+        if @user.save
+            log_in(@user)
```
Comment (nazmy) [inline app/controllers/users_controller.rb]: Good flow. Automatically logging in the user and redirect to new business page 👍

File: app/models/user.rb [hunk 2]
```diff
@@ -1,10 +1,12 @@
 class User < ActiveRecord::Base
+    has_secure_password
 
-  has_secure_password
+    has_one :business
 
-  has_one :business
+    VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
```
Comment (nazmy) [inline app/models/user.rb]: Good 👍.

File: test/fixtures/users.yml [hunk 3]
```diff
@@ -4,8 +4,14 @@
 # model remove the '{}' from the fixture names and add the columns immediately
 # below each fixture, per the syntax in the comments below
 #
-one: {}
-# column: value
-#
-two: {}
-#  column: value
+user1:
```
Comment (nazmy) [inline test/fixtures/users.yml]: Good practice.  Using fixtures to prepare the test data.

**Comment by nazmy** [comment]
Hi Kurnia,

Thanks for your work and PR. No worries, MOKA-0004 is an optional item. 

We have reviewed your submission. We will update you again if you're shortlisted. 

In the meantime, can please help to answer below questions:
1. Do you think there is something that we can improve on the data/entity relation? If so, please briefly explain your answer. 
2. Let's assume you're going to work on item Moka-004. What will you do to implement better authentication and authorisation? Please elaborate your answer. 


Cheers,
Nazmy

**Comment by kurniaw** [comment]
Hi Nazmy,

Thank you for review my submission. Below are my answer of your questions:
1. Yes, with assumption only User, Business and Item model defined.
- To complete audit trail purpose, adding new attributes "t.integer created_by", "t.integer updated_by" on every models defined.
- Change User-Business relation into has_many, so it could accommodate expansion for the customer business.
- On User model itself, probably could add few more attributes, like 
"t.string status": ["active", "disabled"] - for security reason
"t.string role", ["user", "owner", "admin"] - for expansion, probably business owner could invite sub-users to operate their business.
"t.integer fail_login_count", counter number for failed login.
"t.string timezone", for display any datetime field.
"t.datetime last_login_at". just to check last time user login.

2. Yes, i would suggest this additional steps to secure authentication:  
- Firstly implement limit login attempts for block un-authorize login attempt.
- additional captcha verification. additional layer to prevent brute force attack using automated software.
- move authenticate users session into database, so become much more visible & easy to manage. 

Looking forward for update news from Moka.
Thank you.

Regards,
Kurnia W

**Comment by nazmy** [comment]
Hi Kurnia,

Thanks for your answers. I'm fully agree with the explanations. 
Thanks for going through the interview process. We will update you again. 

Regards,
Nazmy

**Pull request closed without merging**